### ISSUE TITLE
[Feature] Embed Subtitles instead of burning to enable end-player subtitle selection

### ISSUE THREAD

**Issue by lenaxia** [ErsatzTV/ErsatzTV#1559]
Currently all subtitles are burned when selected, but the only way to select subtitles for a channel or schedule is by preferred language. This causes problems when there are multiple tracks with the same language, e.g. directors commentary or intro/outro, which can get preferentially selected over the main subtitle track. 

Instead of burning subtitles into the stream, ffmpeg, and m3u8 can support multiple subtitle tracks being streamed, allowing the end user to select their subtitle. This could also be done with audio tracks too, but would be a multiple order of magnitude increase in bandwidth relative to multiple subtitle tracks.

**Proposal**

Currently, there are several subtitle modes, including "Any" which picks a subtitle track that corresponds to the preferred language and burns it into the stream. We propose adding a new mode called "All (Embed)" which does not burn subtitles, instead, it generates .ass files for all subtitles (as opposed to use the preferred language), and then embeds all of them into the ffmpeg stream using `-map 0:s`, or individually mapping each stream. 

To limit overhead, a user could set a list of preferred languages, and only those languages would have their subtitles embedded into the stream.

**Partial Changes based on initial investigation**

Add a new mode
```
public enum ChannelSubtitleMode
{
    None,
    Forced,
    Default,
    Any,
    All // new mode to include all subtitles
}
```

Return all subtitles if mode is set to All
```
public async Task<List<MediaStream>> SelectSubtitleStream(Channel channel, MediaVersion version)
{
    // existing code ...

    if (channel.SubtitleMode == ChannelSubtitleMode.All)
    {
        return subtitleStreams; // return all subtitle streams
    }

    // existing switch statement ...

    return None; // no change here
}
```

Enable SubtitleBuilder to build multiple subtitle tracks:
```
public class SubtitleBuilder
{
    // Existing members of SubtitleBuilder...

    private Option<List<SubtitleTrack>> _subtitleTracks;

    // Existing constructor of SubtitleBuilder...

    public SubtitleBuilder WithSubtitleTracks(List<SubtitleTrack> subtitleTracks)
    {
        _subtitleTracks = subtitleTracks;
        return this;
    }

    public async Task<List<string>> BuildAllFiles()
    {
        List<string> files = new();
        foreach (var subtitleTrack in _subtitleTracks.GetValueOrDefault())
        {
            string fileName = await BuildFile(subtitleTrack);
            files.Add(fileName);
        }
        return files;
    }
```

Individually map each new subtitle stream:
```
    public FFmpegProcessBuilder WithAllSubtitles(IEnumerable<string> subtitlePaths)
    {
        foreach (var subtitlePath in subtitlePaths)
        {
            _arguments.Add(\"-i\");
            _arguments.Add(subtitlePath);
        }
        return this;
    }

    public Process Build()
    {
        var startInfo = new ProcessStartInfo
        {
            FileName = _ffmpegPath,
            RedirectStandardOutput = true,
            RedirectStandardError = false,
            UseShellExecute = false,
            CreateNoWindow = true,
            StandardOutputEncoding = Encoding.UTF8
        };

        startInfo.ArgumentList.Add(\"-nostdin\");
        foreach (string argument in _arguments)
        {
            startInfo.ArgumentList.Add(argument);
        }

        // Logic to map each subtitle stream
        for (int i = 0; i < subtitlePaths.Count(); i++)
        {
            _arguments.Add(\"-map\");
            _arguments.Add($"{i + 1}:s\");
        }

        return new Process
        {
            StartInfo = startInfo
        };
    }
```

Overall I'm not super familiar with the codebase, but this is the basics that I could identify off of a cursory investigation. Pipelines would likely need updating too.

**Comment by Znuff** [comment]
I've looked into this and it's not exactly such an easy task as you THINK it is.

The problem lays in container support for subtitles. 

MPEGTS doesn't do text based subtitles, you'll need to use `dvb_sub`.

`dvb_sub` is an image-based (bitmap) subtitle format, and `ffmpeg` can not convert from TEXT to BITMAP. So you can not just embed a SRT file in MPEGTS and pray. 

This is the same thing when using HLS for streaming, if you're also using TS. 

Fragmented MP4 is also in the same boat, as ffmpeg will usually write the VTT track in a separate file (outside of the fragmented mp4 files). 

Basically, you need to understand that this is "live" streaming, and it's not the same thing as doing things statically on disk.
